pkg/prometheus/operator.go

```diff
@@ -1196,24 +1196,23 @@ func (c *Operator) loadBasicAuthSecrets(
 ) (map[string]BasicAuthCredentials, error) {
 
 	sMonSecretMap := make(map[string]*v1.SecretList)
-	for _, mon := range mons {
-		smNamespace := mon.Namespace
-		if sMonSecretMap[smNamespace] == nil {
-			msClient := c.kclient.CoreV1().Secrets(smNamespace)
-			listSecrets, err := msClient.List(metav1.ListOptions{})
-
-			if err != nil {
-				return nil, errors.Wrapf(err, "failed to retrieve secrets in namespace '%v' for servicemonitor '%v'", smNamespace, mon.Name)
-			}
-			sMonSecretMap[smNamespace] = listSecrets
-		}
-	}
-
 	secrets := map[string]BasicAuthCredentials{}
 	for _, mon := range mons {
 		for i, ep := range mon.Spec.Endpoints {
 			if ep.BasicAuth != nil {
-				credentials, err := loadBasicAuthSecret(ep.BasicAuth, sMonSecretMap[mon.Namespace])
+				smNamespace := mon.Namespace
+				// Load namespace's secrets if not previously fetched
+				if sMonSecretMap[smNamespace] == nil {
+					msClient := c.kclient.CoreV1().Secrets(smNamespace)
+					listSecrets, err := msClient.List(metav1.ListOptions{})
```

Comment:
one problem I see with this current patch is that if a single ServiceMonitor has multiple endpoints (which they often do), we will be listing all the secrets for the same namespace multiple times when there is no need. We should either keep the slice of secrets in a variable in the in the scope of the `mon` for loop or always list from the object cache so we are not making list calls to the API every time.

Reply:
could you explain why that is? As we iterate through the endpoints of a single ServiceMonitor, the first time we notice the namespace hasn't been queried for secrets, fetch the secrets, and then store in the smonSecretMap on line 1212. The next endpoint iteration should notice the cached secrets

Reply:
As I commented and Frederic reiterated, we should only list the secrets from the object cache rather than directly querying the API

Reply:
got it, I see what you mean now

Reply:
IMHO `List` permission is excessive, SMon references concrete object names, so prometheus operators should  `c.kclient.CoreV1().Secrets(smNamespace).Get()` them.

Authors of SMon CR can then provide rolebinding for promether-operator to read  just given secrets without allowing `List`ing all secrets in the namespace

Reply:
Another take on the same problem: https://github.com/coreos/prometheus-operator/pull/2202